cobib.1
.TH COBIB 1 2025-02-08 v5.3.0
.SH NAME
coBib \- Console-based Bibliography Management
.SH SYNOPSIS
.B cobib
[\fB\-\-version\fR]
[\fB\-h\fR|\fB\-\-help\fR]
[\fB\-v\fR|\fB\-\-verbose\fR]
[\fB\-c\fR|\fB\-\-config\fR \fI<path>\fR]
\fB<subcommand>\fR [\fI<args>\fR]
.SH DESCRIPTION
coBib is a console-based bibliography manager written in Python.
It maintains a plain-text database of literature data in YAML format at
\fI$HOME/.local/share/cobib/literature.yaml\fR.
.PP
coBib provides a variety of \fBSUBCOMMANDS\fR through which you may
investigate and manage the database.
For easy of use it also comes with a textual-based \fBTUI\fR which is
automatically started when no other subcommand is found (more information is
provided in that section).
.PP
coBib also provides some "meta"-commands. These are expained in the \fBSHELL
UTILITIES\fR section.
.SH OPTIONS
.TP
.BR \-\-version
Prints the version information and exits.
.TP
.BR \-h ", " \-\-help
Prints a help message and exits.
.TP
.BR \-v ", " \-\-verbose
Increases the verbosity level of the logging. This option may be provided up to
two times (increasing the logging to \fIinfo\fR and \fIdebug\fR, respectively).
By default, the verbosity of coBib's CLI is set to \fIwarning\fR but if the TUI
is started, logging will be increased to \fIinfo\fR and redirected to
\fIconfig.logging.logfile\fR.
.TP
.BR \-p ", " \-\-porcelain
Switches the output that will be printed to the terminal to "porcelain" mode.
This is meant to be useful for parsing and testing purposes.
.TP
.BR \-c ", " \-\-config " " \fI<path>\fR
Run with an alternate configuration file at \fI<path>\fR.
.TP
.BR \-l ", " \-\-log " " \fI<path>\fR
Run with an alternate log file at \fI<path>\fR.
.SH SUBCOMMANDS
All subcommands listed below also provide the \fI\-h\fR and \fI\-\-help\fR
options which provide additional information for each subcommand.
.TP
.B cobib init \fI<args>\fR
Initializes the literature database.
The path to this file can be configured in the \fIDATABASE\fR section of the
configuration file (see also \fBCONFIGURATION\fR).
The only available argument is:
.PP
.in +8n
.BR \-g ", " \-\-git
.in +4n
Initializes coBib's integrated history tracking through git.
This command will preserve any existing database file, but it is a good idea to
make a backup before doing this, just in case.
Note, that in order to use this functionality you must also enable the
\fIDATABASE/git\fR configuration option (see also \fBCONFIGURATION\fR).
Also be sure to at least set a \fIname\fR and \fIemail\fR in the git config!
.TP
.B cobib git \fI<args>\fR ...
.in +4n
Passes through to \fIgit\fR to run operations on the underlying database.
All arguments provided after \fIgit\fR are forwarded without being changed.
.TP
.B cobib add \fI<args>\fR ...
Adds a new entry to the database.
The positional arguments may be used to provide \fItags\fR to associate with the
newly added entries.
The \fI<args>\fR may be any of the following:
.PP
.in +8n
.BR \-a ", " \-\-arxiv " " \fI<arxiv\ id>\fR
.in +4n
Adds an entry specified by the \fIarXiv\fR id.
.PP
.in +8n
.BR \-b ", " \-\-bibtex " " \fI<path>\fR
.in +4n
Adds the bibliography data from the \fIBibLaTex\fR file at the provided path.
.PP
.in +8n
.BR \-d ", " \-\-doi " " \fI<doi>\fR
.in +4n
Adds an entry specified by the \fIDOI\fR.
.PP
.in +8n
.BR \-i ", " \-\-isbn " " \fI<ISBN>\fR
.in +4n
Adds an entry specified by the \fIISBN\fR.
.PP
.in +8n
.BR \-\-url " " \fI<URL>\fR
.in +4n
Attempts to extract an entry from the \fI<URL>\fR.
.PP
.in +8n
.BR \-y ", " \-\-yaml " " \fI<path>\fR
.in +4n
Adds the bibliography data from the \fIYAML\fR file at the provided path.
The only supported YAML format is that of a coBib database file.
.PP
.in +8n
.BR \-l ", " \-\-label  " " \fI<label>\fR
.in +4n
Store the newly added entry under the specified \fIlabel\fR.
.PP
.in +8n
.BR \-f ", " \-\-file " " \fI<path>\fR
.in +4n
Associate the newly added entry with the \fIfile\fR at the provided path.
As of version 2.5 you can specify multiple files, too.
.PP
.in +8n
.BR \-p ", " \-\-path " " \fI<path>\fR
.in +4n
Overwrites the location for the automatically downloaded file. If not specified,
this setting defaults to \fIconfig.utils.file_downloader.default_location\fR.
.PP
.in +8n
.BR \-\-disambiguation " " \fI<reply>\fR
.in +4n
Sets the reply for the interactive prompt if an addition needs to be
disambiguated. The possible values for \fI<reply>\fR are the following:
.in +4n
.IR keep
to keep the existing entry
.in +0n
.IR replace
to replace the existing entry
.in +0n
.IR update
to update the existing entry
.in +0n
.IR disambiguate
to disambiguate the new label from the existing one by adding a label suffix
(see also \fIconfig.database.format.label_suffix\fR)
.PP
.in +8n
.BR \-\-skip\-download
.in +4n
Skips attempting to automatically download an associated file for the entry.
This takes precedence over the \fIconfig.commands.add.skip_download\fR setting.
.PP
.in +8n
.BR \-\-force\-download
.in +4n
Forces attempting to automatically download an associated file for the entry.
This takes precedence over the \fIconfig.commands.add.skip_download\fR setting.
.TP
.B cobib delete \fI<label>\fR
Deletes the entry with the given \fIlabel\fR.
.PP
.in +8n
.BR \-\-preserve\-files
.in +4n
If specified, associated files will be preserved and \fInot\fR deleted.
This takes precedence over the \fIconfig.commands.delete.preserve_files\fR
setting.
.PP
.in +8n
.BR \-\-no\-preserve\-files
.in +4n
If specified, associated files will \fInot\fR be preserved.
This takes precedence over the \fIconfig.commands.delete.preserve_files\fR
setting.
.TP
.B cobib edit \fI<label>\fR
Opens the entry with the given \fIlabel\fR in an external editor.
The entry is copied verbatim in \fIYAML\fR format from and to the database file.
The editor respects the \fI$EDITOR\fR environment variable unless overwritten by
\fIconfig.commands.edit.editor\fR. It will fallback to \fIvim\fR if neither is
configured.
.PP
.in +8n
.BR \-a ", " \-\-add " " \fI<path>\fR
.in +4n
Allows editing a non-existing label to manually add it to the database.
.PP
.in +8n
.BR \-\-preserve\-files
.in +4n
If specified, associated files will be preserved and \fInot\fR be renamed if the
edit happens to rename the entry. This takes precedence over the
\fIconfig.commands.edit.preserve_files\fR setting.
.PP
.in +8n
.BR \-\-no\-preserve\-files
.in +4n
If specified, associated files will \fInot\fR be preserved and be renamed if the
edit happens to rename the entry. This takes precedence over the
\fIconfig.commands.edit.preserve_files\fR setting.
.TP
.B cobib modify \fI<modification>\fR \fI<args>\fR ...
Applies a modification to multiple entries at once.
The positional arguments may be used to provide \fBFILTERS\fR which the entries
must match in order to be modified \fIor\fR to provide a list of labels of the
entries which are to be modified (this requires the \fI-s\fR flag to be set).
The \fI<args>\fR may be any of the following:
.PP
.in +8n
.BR \fI<modification>\fR
.in +4n
The modification must be provided in the format \fI<field>:<value>\fR and will
set the field of all selected entries to the given value.
As of v3.2.0 the \fI<value>\fR is interpreted as an "f"-string. This means you
can even use placeholder variables and perform simple operations on them. The
available variables depend on the entry which you are modifying as they are
inferred from its stored data. For more information on "f"-strings refer to
section 2.4.3 of https://docs.python.org/3/reference/lexical_analysis.html
.PP
.in +8n
.BR \-\-dry
.in +4n
When this flag is given, the modify command runs in \fIdry\fR mode. This means,
the applied modifications are printed to stdout rather than applied directly.
This allows easy prototyping of modifications to prevent errors during large
bulk modifications.
.PP
.in +8n
.BR \-a ", " \-\-add
.in +4n
Specifying this flag will add the modification's value to the specified field of
the entry rather than overwriting it. In doing so, strings will be concatenated
with\fBOUT\fR any spaces, lists will be appended to, numbers will be added, and
anything else will be converted to a string field.
If the field did not exist previously, it will simply be initialized with the
new value.
.PP
.in +8n
.BR \-r ", " \-\-remove
.in +4n
Specifying this flag will remove the modification's value from the specified
field of the entry rather than overwriting it. In doing so, items will be
removed from lists, numbers will be subtracted, and anything else will be left
unchanged. However, as of v5.3.0 you can combine this with an empty
\fI<value>\fR to outright remove a field from the modified entries.
.PP
.in +8n
.BR \-s ", " \-\-selection
.in +4n
This boolean flag enables the \fIselection\fR mode in which the positional args
are interpreted as a list of labels which are to be exported. The name for this
argument is a result of the TUI's selection interface.
.PP
.in +8n
.BR \-\-preserve\-files
.in +4n
If specified, associated files will be preserved and \fInot\fR be renamed if
the modification happens to rename the entry. This takes precedence over the
\fIconfig.commands.modify.preserve_files\fR setting.
.PP
.in +8n
.BR \-\-no\-preserve\-files
.in +4n
If specified, associated files will \fInot\fR be preserved and be renamed if
the modification happens to rename the entry. This takes precedence over the
\fIconfig.commands.modify.preserve_files\fR setting.
.TP
.B cobib note \fI<label>\fR \fI<action>\fR
Applies an action to the note of the provided entry. The available \fIactions\fR
are:
.PP
.in +8n
.BR show
.in +4n
Shows the note's contents.
.PP
.in +8n
.BR edit
.in +4n
Opens the note file in the \fIconfig.commands.edit.editor\fR for editing.
.PP
.in +8n
.BR delete
.in +4n
Deletes the note.
.TP
.B cobib review \fI<field>\fR \fI<args>\fR ...
Performs an interactive review of your database.
You can optionally review one or more specific fields of the entries by
providing the field names as arguments to the command.
The positional arguments may be used to provide \fBFILTERS\fR which the entries
must match in order to be reviewed \fIor\fR to provide a list of labels of the
entries which are to be reviewed (this requires the \fI-s\fR flag to be set).
For every entry that is being reviewed, multiple actions are possible. These
include:
.TP
.in +8n
.BR edit
.in +4n
Opens the current entry in an external editor.
.TP
.in +8n
.BR done
.in +4n
Marks the current entry as fully reviewed.
.TP
.in +8n
.BR skip
.in +4n
Skips the current entry for this review.
.TP
.in +8n
.BR context
.in +4n
Requests additional context and updates the preview to contain all the
information stored in the entry. This action is only available when no context
has been requested yet and the previewed fields are being filtered in the first
place.
.TP
.in +8n
.BR finish
.in +4n
Finishes the review early (even if some entries have not been shown yet). See
\fI\-\-resume\fR for how to resume a previously started review process.
.TP
.in +8n
.BR inline <field>
.in +4n
Edits a specific field in-line. When reviewing a single field, you can simply
type \fIinline\fR. When reviewing more than one, you need to specify which field
to edit like so \fIinline <field>\fR. This is a highly experimental feature so
please be aware of bugs and report any issues or suggestions online:
https://gitlab.com/cobib/cobib/-/issues/new
.TP
.in +8n
The \fI<args>\fR to the \fBreview\fR command may be any of the following:
.PP
.in +8n
.BR \-c ", " \-\-context
.in +4n
.PP
Enforces that context will be provided for every entry.
.in +8n
.BR \-r ", " \-\-resume " " \fI<SHA>\fR
.in +4n
If you finished a review early (see `finish` above) you can use this argument to
continue where you left. This option takes a git commit identifier as its
argument so you should either find the SHA of the auto-commit of the review that
you would like to continue (see also the \fIgit\fR command), or if you know that
it is (for example) the last thing that you did, you can simply use \fRHEAD\fR.
.PP
.in +8n
.BR \-d ", " \-\-done " " \fI<labels>\fR
.in +4n
Can supply a list of entry labels which will be considered as already reviewed
and therefore skipped. Thus, this has a higher precedence than the provided
\fBFILTERS\fR and can be used to resume a previously started review process. If
you have the \fIgit\fR integration enabled, you will likely not use this
argument manually. Instead, refer to the \fI\-\-resume\fR argument.
.PP
.in +8n
.BR \-s ", " \-\-selection
.in +4n
This boolean flag enables the \fIselection\fR mode in which the positional args
are interpreted as a list of labels which are to be exported. The name for this
argument is a result of the TUI's selection interface.
.TP
.B cobib undo \fI<args>\fR
If you enabled the git-integration of coBib (available since v2.6.0) you can
undo the changes done to your database file by commands such as add, edit and
delete. See also \fIDATABASE/git\fR in the \fBCONFIGURATION\fR section for more
information.
.PP
.in +8n
.BR \-f ", " \-\-force
.in +4n
Overwrites the check for an auto-committed change. Thus, the undo command will
now undo the last commit with a message that does not start with "Undo".
.TP
.B cobib redo
If you enabled the git-integration of coBib (available since v2.6.0) you can
reapply the last undone changes (see above). See also \fIDATABASE/git\fR in the
\fBCONFIGURATION\fR section for more information.
.TP
.B cobib open \fI<label>\fR
Opens any associated \fIfile\fR of the entry with the given \fIlabel\fR.
If multiple files are associated with the entry, the user can choose which
file(s) to open through an interactive menu.
.PP
.in +8n
.BR \-f ", " \-\-field " " \fI<choice>\fR
.in +4n
Specifies the field type to open. This bypasses the interactive prompt if
multiple actionable fields are found. The choice can be either \fIall\fR or any
of the values configured in \fIconfig.commands.open.fields\fR.
.TP
.B cobib show \fI<label>\fR
Prints the entry with the given \fIlabel\fR in \fIBibLaTex\fR format to stdout.
.TP
.B cobib list \fI<args>\fR
Lists all entries of the database in a basic table format to stdout which match
the specified \fBFILTERS\fR (more information is provided in that section).
Additionally, the following \fI<args>\fR are also allowed:
.PP
.in +8n
.BR \-s ", " \-\-sort " " \fI<field>\fI
.in +4n
Specify the entry field to use as the \fIsorting column\fR of the table.
.PP
.in +8n
.BR \-r ", " \-\-reverse
.in +4n
Reverses the sorting order.
.PP
.in +8n
.BR \-l ", " \-\-limit " " \fI<int>\fI
.in +4n
Limits the number of displayed results to the specified number.
.PP
.in +8n
.BR \-i ", " \-\-ignore\-case
.in +4n
Makes the entry matching case-insensitive.
This takes precedence over the \fIconfig.commands.list_.ignore_case\fR setting.
.PP
.in +8n
.BR \-I ", " \-\-no\-ignore\-case
.in +4n
Makes the entry matching case-sensitive.
This takes precedence over the \fIconfig.commands.list_.ignore_case\fR setting.
.PP
.in +8n
.BR \-\-decode\-latex
.in +4n
Makes the entry matching decode all LaTeX sequences.
This takes precedence over the \fIconfig.commands.list_.decode_latex\fR setting.
.PP
.in +8n
.BR \-\-no\-decode\-latex
.in +4n
Makes the entry matching preserve all LaTeX sequences.
This takes precedence over the \fIconfig.commands.list_.decode_latex\fR setting.
.PP
.in +8n
.BR \-\-decode\-unicode
.in +4n
Makes the entry matching decode all Unicode characters.
This takes precedence over the \fIconfig.commands.list_.decode_unicode\fR
setting.
.PP
.in +8n
.BR \-\-no\-decode\-unicode
.in +4n
Makes the entry matching preserve all Unicode characters.
This takes precedence over the \fIconfig.commands.list_.decode_unicode\fR
setting.
.PP
.in +8n
.BR \-z ", " \-\-fuzziness " " \fI<int>\fI
.in +4n
Specifies how many fuzzy errors to allow during entry matching.
The default value is 0 but can be configured via
\fIconfig.commands.list_.fuzziness\fR.
.PP
.in +8n
.BR \-x ", " \-\-or
.in +4n
Concatenate the filters using logical \fIOR\fR rather than the default
\fIAND\fR.
.TP
.B cobib search \fI<query>\fR \fI<args>\fR ...
Searches the database recursively (i.e. including any associated files) for the
specified queries (if multiple are given, these will be searched independently).
The positional arguments may be used to provide \fBFILTERS\fR which the entries
must match in order to be included in the export.
Additionally, the following \fI<args>\fR are also allowed:
.PP
.in +8n
.BR \-c ", " \-\-context " " \fI<int>\fI
.in +4n
Specify the number of context lines to provide for each match.
The default value is 1 but can be configured via
\fIconfig.commands.search.context\fR.
.PP
.in +8n
.BR \-i ", " \-\-ignore\-case
.in +4n
Makes the search case-insensitive.
This takes precedence over the \fIconfig.commands.search.ignore_case\fR setting.
.PP
.in +8n
.BR \-I ", " \-\-no\-ignore\-case
.in +4n
Makes the search case-insensitive.
This takes precedence over the \fIconfig.commands.search.ignore_case\fR setting.
.PP
.in +8n
.BR \-l ", " \-\-decode\-latex
.in +4n
Makes the search decode all LaTeX sequences.
This takes precedence over the \fIconfig.commands.search.decode_latex\fR
setting.
.PP
.in +8n
.BR \-L ", " \-\-no\-decode\-latex
.in +4n
Makes the search preserve all LaTeX sequences.
This takes precedence over the \fIconfig.commands.search.decode_latex\fR
setting.
.PP
.in +8n
.BR \-u ", " \-\-decode\-unicode
.in +4n
Makes the search decode all Unicode characters.
This takes precedence over the \fIconfig.commands.search.decode_unicode\fR
setting.
.PP
.in +8n
.BR \-U ", " \-\-no\-decode\-unicode
.in +4n
Makes the search preserve all Unicode characters.
This takes precedence over the \fIconfig.commands.search.decode_unicode\fR
setting.
.PP
.in +8n
.BR \-z ", " \-\-fuzziness " " \fI<int>\fI
.in +4n
Specifies how many fuzzy errors to allow during search.
The default value is 0 but can be configured via
\fIconfig.commands.search.fuzziness\fR.
.PP
.in +8n
.BR \-\-skip\-files
.in +4n
Skips searching the associated files.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [5.3.0] - 2025-02-08

Pypi: https://pypi.org/project/cobib/5.3.0/

### Added

- the ability to remove fields entirely via the `modify` command (#142,!299)
- a hint in the documentation how to filter for (non-)existence of fields (#141)
- the coBib command `prompt` to the TUI's command palette (!310)

### Changed
- HTTP requests are now done inside of `requests.Session` instances (!233)

### Fixed
- the `Escape` key during an input prompt triggered after a `Confirm` prompt (e.g. a cancelled `quit`)
- the behavior of the `Home` and `End` keybindings in the TUI's search view (!310)
- the ability to `quit` the application from the command palette (!310)


## [5.2.3] - 2024-12-15

Pypi: https://pypi.org/project/cobib/5.2.3/

### Fixed

- the interactive handling of notes inside the TUI (#148,!288)


## [5.2.2] - 2024-12-06

Pypi: https://pypi.org/project/cobib/5.2.2/

### **Breaking Change**

- The `config.theme.dark` and `config.theme.design` settings have been replaced
  by the new `config.theme.theme` setting, which can be either the name of one
  of textual's builtin themes, or a custom `textual.theme.Theme` instance
  (#147,!279). Due to how recent these settings have been introduced (as part of
  5.2.0), I am opting for a quick removal in this bugfix release rather than a
  proper version bump (in the sense that they have likely not been adopted, yet)
  and the severness of the bug that this is causing outweighs the needs here.

### Fixed

- compatibility with `textual>=0.86`


## [5.2.1] - 2024-10-30

Pypi: https://pypi.org/project/cobib/5.2.1/

### Fixed

- compatibility with `textual>=0.80`


## [5.2.0] - 2024-10-30

Pypi: https://pypi.org/project/cobib/5.2.0/

### Prelude

With the addition of the new `note` command, the database entry fields called
`"note"` have obtained a special meaning. Use the `lint` command to check
whether your database is affected by this. If it is, you should rename those
fields.

### Added
- new configuration settings to customize the appearance of coBib (!222):
  - `cobib.config.theme.dark`: to specify whether a dark or light theme is used
  - `cobib.config.theme.design`: to customize the colors themselves
  - the following settings to customize the appearance of `rich.Syntax` displays:
    - `cobib.config.theme.syntax.theme`
    - `cobib.config.theme.syntax.background_color`
    - `cobib.config.theme.syntax.line_numbers`
- the `note` command (#139,!206).
  Please refer to its online documentation or the man-page for more details on
  how to use it.
  This new command comes with:
  - the `cobib.config.commands.note.default_filetype` setting
  - the `PreNoteCommand` and `PostNoteCommand` event hooks
- (DEV) the `add_files` argument to the `BaseCommand.git` function
- Python 3.13 is now officially tested and supported
- new configuration settings for the search command (#146,!269)
  - For dealing with skipping/including associated files during searches:
    - `cobib.config.commands.search.skip_files` sets the default
    - `--skip-files` overwrites this at runtime (this is not actually new)
    - `--include-files` overwrites this at runtime (the opposite)
  - For dealing with skipping/including associated notes during searches:
    - `cobib.config.commands.search.skip_notes` sets the default
    - `--skip-notes` overwrites this at runtime (this is not actually new)
    - `--include-notes` overwrites this at runtime (the opposite)

### Changed
- the keybinding to recursively toggle all folds in the TUI's search result view
  was changed from `ENTER` to `BACKSPACE` (!218)
- (DEV) `cobib.ui.components.entry_points` was moved to `cobib.utils.entry_points`
- replaced the custom help screen with textual's new builtin help panel (!240)

### Fixed
- the behavior of the `Home` and `End` keybindings in the TUI's list view (!217)
- gracefully handles Zotero imported entries without a `journal` field (#145)


## [5.1.3] - 2024-08-23

Pypi: https://pypi.org/project/cobib/5.1.3/

### Fixed
- the behavior of the `Home` and `End` keybindings in the TUI's list view (!217)


## [5.1.2] - 2024-06-23

Pypi: https://pypi.org/project/cobib/5.1.2/

### Fixed
- support database locations in nested folders of a git repository (!204)


## [5.1.1] - 2024-05-28

Pypi: https://pypi.org/project/cobib/5.1.1/

### Fixed

- when printing the changelog, do not remove list indents


## [5.1.0] - 2024-05-28

Pypi: https://pypi.org/project/cobib/5.1.0/

### Prelude

Check out my blog post to learn more about the features of this new release:
https://mrossinek.gitlab.io/programming/cobib-becomes-fuzzy/

### Added
- approximate (or fuzzy) filter matching and search functionality (#107,#130,!177)
  - the `list` and `search` commands now support the following features to
    perform approximate filter matching and searching, respectively:
    - LaTeX sequences can be decoded to Unicode characters:
      - using `--decode-latex` from the command-line
      - setting `config.commands.list_.decode_latex = True`
      - setting `config.commands.search.decode_latex = True`
    - Unicode characters can be converted to a close ASCII equivalent:
      - using `--decode-unicode` from the command-line
      - setting `config.commands.list_.decode_unicode = True`
      - setting `config.commands.search.decode_unicode = True`
    - a number of fuzzy errors can be set (this requires the optional dependency
      [`regex`](https://pypi.org/project/regex/) to be installed):
      - using `--fuzziness <int>` from the command-line
      - setting `config.commands.list_.fuzzines` to some integer
      - setting `config.commands.search.fuzzines` to some integer
- (DEV) the following method arguments have been converted to be accepted only

cobib/cobib.py
"""CoBib main module"""

# IMPORTS
# standard
import argparse
import os
import sys
import tempfile
import textwrap
from collections import OrderedDict, defaultdict
from operator import itemgetter
from pathlib import Path
from subprocess import Popen
from zipfile import ZipFile
# third-party
import tabulate
# relative
from .parser import Entry
from .config import CONFIG


# ARGUMENT FUNCTIONS
def init_(args):  # pylint: disable=unused-argument
    """initialize database

    Initializes the yaml database file at the configured location.
    """
    conf_database = dict(CONFIG['DATABASE'])
    file = os.path.expanduser(conf_database['file'])
    open(file, 'w').close()


def list_(args, out=sys.stdout):  # pylint: disable=too-many-branches,too-many-locals
    """list entries

    By default, all entries of the database are listed.
    This output will be filterable in the future by providing values for any
    set of table keys.
    """
    if '--' in args:
        args.remove('--')
    parser = argparse.ArgumentParser(prog="list", description="List subcommand parser.",
                                     prefix_chars='+-')
    parser.add_argument('-x', '--or', dest='OR', action='store_true',
                        help="concatenate filters with OR instead of AND")
    parser.add_argument('-l', '--long', action='store_true',
                        help="print table in long format (i.e. wrap long lines rather than shorten")
    parser.add_argument('-s', '--sort', help="specify column along which to sort the list")
    parser.add_argument('-r', '--reverse', action='store_true',
                        help="reverses the sorting order")
    bib_data = _read_database()
    unique_keys = set()
    for entry in bib_data.values():
        unique_keys.update(entry.data.keys())
    for key in sorted(unique_keys):
        parser.add_argument('++'+key, type=str, action='append',
                            help="include elements with matching "+key)
        parser.add_argument('--'+key, type=str, action='append',
                            help="exclude elements with matching "+key)
    largs = parser.parse_args(args)
    _filter = defaultdict(list)
    for key, val in largs.__dict__.items():
        if key in ['OR', 'long', 'sort', 'reverse'] or val is None:
            continue
        if not isinstance(val, list):
            val = [val]
        for i in val:
            for idx, obj in enumerate(args):
                if i == obj:
                    _filter[tuple([key, args[idx-1][0] == '+'])].append(i)
                    break
    columns = ['ID', 'title']
    if largs.sort and largs.sort not in columns:
        columns.append(largs.sort)
    columns.extend([arg[0] for arg in _filter.keys() if arg[0] not in columns])
    labels = []
    table = []
    for key, entry in bib_data.items():
        if entry.matches(_filter, largs.OR):
            labels.append(key)
            table.append([entry.data.get(c, None) for c in columns])
            if largs.long:
                table[-1][1] = textwrap.fill(table[-1][1], width=80)
            else:
                table[-1][1] = textwrap.shorten(table[-1][1], 80, placeholder='...')
    if largs.sort:
        table = sorted(table, key=itemgetter(columns.index(largs.sort)), reverse=largs.reverse)
    print(tabulate.tabulate(table, headers=columns), file=out)
    return labels


def show_(args, out=sys.stdout):
    """show entry

    Prints the details of a selected entry in bibtex format to stdout.
    """
    parser = argparse.ArgumentParser(prog="show", description="Show subcommand parser.")
    parser.add_argument("label", type=str, help="label of the entry")
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)
    bib_data = _read_database()
    try:
        entry = bib_data[largs.label]
        entry_str = entry.to_bibtex()
        print(entry_str, file=out)
    except KeyError:
        print("Error: No entry with the label '{}' could be found.".format(largs.label))


def open_(args):
    """open file from entry

    Opens the associated file of an entry with xdg-open.
    """
    parser = argparse.ArgumentParser(prog="open", description="Open subcommand parser.")
    parser.add_argument("label", type=str, help="label of the entry")
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)
    bib_data = _read_database()
    try:
        entry = bib_data[largs.label]
        if 'file' not in entry.data.keys() or entry.data['file'] is None:
            print("Error: There is no file associated with this entry.")
            sys.exit(1)
        try:
            Popen(["xdg-open", entry.data['file']], stdin=None, stdout=None, stderr=None,
                  close_fds=True, shell=False)
        except FileNotFoundError:
            try:
                Popen(["open", entry.data['file']], stdin=None, stdout=None, stderr=None,
                      close_fds=True, shell=False)
            except FileNotFoundError:
                pass
    except KeyError:
        print("Error: No entry with the label '{}' could be found.".format(largs.label))


def add_(args):
    """add new entry

    Adds new entries to the database.
    """
    parser = argparse.ArgumentParser(prog="add", description="Add subcommand parser.")
    parser.add_argument("-l", "--label", type=str,
                        help="the label for the new database entry")
    parser.add_argument("-f", "--file", type=str,
                        help="a file associated with this entry")
    group_add = parser.add_mutually_exclusive_group()
    group_add.add_argument("-a", "--arxiv", type=str,
                           help="arXiv ID of the new references")
    group_add.add_argument("-b", "--bibtex", type=argparse.FileType('r'),
                           help="BibTeX bibliographic data")
    group_add.add_argument("-d", "--doi", type=str,
                           help="DOI of the new references")
    parser.add_argument("tags", nargs=argparse.REMAINDER)
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)

    new_entries = OrderedDict()

    if largs.bibtex is not None:
        new_entries = Entry.from_bibtex(largs.bibtex)
    if largs.arxiv is not None:
        new_entries = Entry.from_arxiv(largs.arxiv)
    if largs.doi is not None:
        new_entries = Entry.from_doi(largs.doi)

    if largs.file is not None:
        assert len(new_entries.values()) == 1
        for value in new_entries.values():
            value.set_file(largs.file)

    if largs.label is not None:
        assert len(new_entries.values()) == 1
        for value in new_entries.values():
            value.set_label(largs.label)
        new_entries = OrderedDict((largs.label, value) for value in new_entries.values())

    if largs.tags != []:
        assert len(new_entries.values()) == 1
        for value in new_entries.values():
            value.set_tags(largs.tags)

    _write_database(new_entries)


def remove_(args):
    """remove entry

    Removes the entry from the database.
    """
    parser = argparse.ArgumentParser(prog="remove", description="Remove subcommand parser.")
    parser.add_argument("label", type=str, help="label of the entry")
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)
    conf_database = dict(CONFIG['DATABASE'])
    file = os.path.expanduser(conf_database['file'])
    with open(file, 'r') as bib:
        lines = bib.readlines()
    entry_to_be_removed = False
    buffer = []
    for line in lines:
        if line.startswith(largs.label):
            entry_to_be_removed = True
            buffer.pop()
            continue
        if entry_to_be_removed and line.startswith('...'):
            entry_to_be_removed = False
            continue
        if not entry_to_be_removed:
            buffer.append(line)
    with open(file, 'w') as bib:
        for line in buffer:
            bib.write(line)


def edit_(args):
    """edit entry

    Opens an existing entry for manual editing.
    """
    parser = argparse.ArgumentParser(prog="edit", description="Edit subcommand parser.")
    parser.add_argument("label", type=str, help="label of the entry")
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)
    bib_data = _read_database()
    try:
        entry = bib_data[largs.label]
        prv = entry.to_yaml()
    except KeyError:
        print("Error: No entry with the label '{}' could be found.".format(largs.label))
    tmp_file = tempfile.NamedTemporaryFile(mode='w+', suffix='.yaml')
    tmp_file.write(prv)
    tmp_file.flush()
    status = os.system(os.environ['EDITOR'] + ' ' + tmp_file.name)
    assert status == 0
    tmp_file.seek(0, 0)
    nxt = ''.join(tmp_file.readlines()[1:])
    tmp_file.close()
    assert not os.path.exists(tmp_file.name)
    if prv == nxt:
        return
    conf_database = dict(CONFIG['DATABASE'])
    file = os.path.expanduser(conf_database['file'])
    with open(file, 'r') as bib:
        lines = bib.readlines()
    entry_to_be_replaced = False
    with open(file, 'w') as bib:
        for line in lines:
            if line.startswith(largs.label):
                entry_to_be_replaced = True
                continue
            if entry_to_be_replaced and line.startswith('...'):
                entry_to_be_replaced = False
                bib.writelines(nxt)
                continue
            if not entry_to_be_replaced:
                bib.write(line)


def export_(args):
    """export database

    Exports all entries matched by the filter queries (see the list docs).
    Currently supported exporting formats are:
    * bibtex databases
    * zip archives
    """
    parser = argparse.ArgumentParser(prog="export", description="Export subcommand parser.")
    parser.add_argument("-b", "--bibtex", type=argparse.FileType('a'),
                        help="BibTeX output file")
    parser.add_argument("-z", "--zip", type=argparse.FileType('a'),
                        help="zip output file")
    parser.add_argument('list_args', nargs=argparse.REMAINDER)
    if not args:
        parser.print_usage(sys.stderr)
        sys.exit(1)
    largs = parser.parse_args(args)
    bib_data = _read_database()

    if largs.bibtex is None and largs.zip is None:
        return
    if largs.zip is not None:
        largs.zip = ZipFile(largs.zip.name, 'w')
    out = open(os.devnull, 'w')
    labels = list_(largs.list_args, out=out)

    try:
        for label in labels:
            entry = bib_data[label]
            if largs.bibtex is not None:
                entry_str = entry.to_bibtex()
                largs.bibtex.write(entry_str)
            if largs.zip is not None:
                if 'file' in entry.data.keys() and entry.data['file'] is not None:
                    largs.zip.write(entry.data['file'], label+'.pdf')
    except KeyError:
        print("Error: No entry with the label '{}' could be found.".format(largs.label))


# HELPER FUNCTIONS
def _read_database():
    conf_database = dict(CONFIG['DATABASE'])
    file = os.path.expanduser(conf_database['file'])
    try:
        bib_data = Entry.from_yaml(Path(file))
    except AttributeError:
        bib_data = OrderedDict()
    return bib_data


def _write_database(entries):
    bib_data = _read_database()
    new_lines = []
    for label, entry in entries.items():
        if label in bib_data.keys():
            print("Error: label '{}' already exists!".format(label))
            continue
        string = entry.to_yaml()
        reduced = '\n'.join(string.splitlines())
        new_lines.append(reduced)

    conf_database = dict(CONFIG['DATABASE'])
    file = os.path.expanduser(conf_database['file'])
    with open(file, 'a') as bib:
        for line in new_lines:
            bib.write(line+'\n')
